src/amperity/gocd/secret/vault/plugin.clj

```diff
@@ -110,25 +128,83 @@
+                           (keys input-schema))})
+
+
+(defn- input-error-message
+  "If the input field is valid, does nothing, if it's not, returns a string describing the error.
+
+  Params:
+  - `field-key`: the input field you are validating
+  - `field-value`: the inputted value which you wish to verify
+  - `field-label`: the human readable representation of the field key"
+  [field-key field-value]
+  (let [field-model (field-key input-schema)
+        validate-func-errors (remove #((:func %)  field-value)
+                                     (:validate-funcs field-model))]
+    (cond
+      ;; field is required but empty
+      (and (-> field-model :metadata :required) (str/blank? field-value))
+      (str (:label field-model) " is required")
+
+      ;; field is not empty and there is an error
+      (not (or (str/blank? field-value) (empty? validate-func-errors)))
+      (:message (first validate-func-errors))
+
+      :else
+      nil)))
+
+
+(defn- authenticate-client-from-inputs!
+  "Authenticates the Vault Client
+
+  Params:
+  - `client` The Vault Client you wish to authenticate
+  - `inputs` A map containing the user inputted settings for the plugin"
+  [client inputs]
+  (case (keyword (:auth_method inputs))
+    :token
+    (vault/authenticate! client :token (:vault_token inputs))
+
+    (throw (ex-info "Could not recognize user inputted vault auth type"
+                    {:user-input (keyword (:auth_method inputs))}))))
 
 
 ;; This call is expected to validate the user inputs that form a part of
 ;; the secret backend configuration.
-(defmethod handle-request "go.cd.screts.validate"
-  [_ _ data]
-  (letfn [(validate-string
-            [field-key label]
-            (when (str/blank? (get data field-key))
-              {:key     field-key
-               :message (str label " is required")}))]
-    {:response-code    200
-     :response-headers {}
-     :response-body    (into [] (remove nil?) [(validate-string :vault_addr "Vault URL")])}))
+(defmethod handle-request "go.cd.secrets.validate"
+  [client _ data]
+  (let [input-error (fn [field-key]
+                      {:key   field-key
+                       :message (input-error-message field-key (field-key data))})
+        errors-found (filterv :message (map input-error (keys input-schema)))]
+    (if (and (empty? errors-found)
+             ;; Need to Authenticate?
+             (not (and (= (:api-url @client) (:vault_addr data))
+                       (= (:auth-type @client) (:auth_method data))
+                       (= (:client-token @client) (:vault_token data)))))
+      ;; Authenticate Vault client
+      (try
+        (when (not (= (:api-url @client) (:vault_addr data)))
+          (reset! client (vault/new-client (:vault_addr data))))
+        (authenticate-client-from-inputs! @client data)
+        {:response-code 200
+         :response-headers {}
+         :response-body []}
+
+        (catch Exception ex
+          {:response-code 200
```

Comment:
A 200 is surprising for this case. is this supporting something gocd is expecting/ignoring?

Reply:
Yeah, GoCD is expecting 200 on all responses for well-formed requests here: https://plugin-api.gocd.org/current/secrets/#validate-secret-configuration